lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts

```diff
@@ -61,47 +111,307 @@ export class UnifiedCacheManager {
      * @param homeAccountId
      * @param environment
+     * @param credentialType
+     * @param clientId
      * @param realm
+     * @param target
      */
-    getAccount(homeAccountId: string, environment: string, realm: string): AccountEntity {
-        const accountCacheKey: Array<string> = [
+    getCredentialsFilteredBy(
+        homeAccountId?: string,
+        environment?: string,
+        credentialType?: string,
+        clientId?: string,
+        realm?: string,
+        target?: string
+    ): CredentialCache {
+        const matchingCredentials: CredentialCache = {
+            idTokens: {},
+            accessTokens: {},
+            refreshTokens: {},
+        };
+
+        matchingCredentials.idTokens = this.getCredentialsFilteredByInternal(
+            this.getCacheInMemory().idTokens,
             homeAccountId,
             environment,
-            realm
-        ];
+            credentialType,
+            clientId,
+            realm,
+            target
+        ) as IdTokenCache;
 
-        const accountKey = accountCacheKey.join(Separators.CACHE_KEY_SEPARATOR).toLowerCase();
+        matchingCredentials.accessTokens = this.getCredentialsFilteredByInternal(
+            this.getCacheInMemory().accessTokens,
+            homeAccountId,
+            environment,
+            credentialType,
+            clientId,
+            realm,
+            target
+        ) as AccessTokenCache;
+
+        matchingCredentials.refreshTokens = this.getCredentialsFilteredByInternal(
+            this.getCacheInMemory().refreshTokens,
+            homeAccountId,
+            environment,
+            credentialType,
+            clientId,
+            realm,
+            target
+        ) as RefreshTokenCache;
 
-        return this.inMemoryCache.accounts[accountKey] || null;
+        return matchingCredentials;
     }
 
     /**
-     * append credential cache to in memory cache
-     * @param idT: IdTokenEntity
-     * @param at: AccessTokenEntity
-     * @param rt: RefreshTokenEntity
+     * Support function to help match credentials
+     * @param cacheCredentials
+     * @param homeAccountId
+     * @param environment
+     * @param credentialType
+     * @param clientId
+     * @param realm
+     * @param target
      */
-    addCredentialCache(
-        accessToken: AccessTokenEntity,
-        idToken: IdTokenEntity,
-        refreshToken: RefreshTokenEntity
-    ): void {
-        this.inMemoryCache.accessTokens[accessToken.generateCredentialKey()] = accessToken;
-        this.inMemoryCache.idTokens[idToken.generateCredentialKey()] = idToken;
-        this.inMemoryCache.refreshTokens[refreshToken.generateCredentialKey()] = refreshToken;
+    private getCredentialsFilteredByInternal(
+        cacheCredentials: object,
+        homeAccountId?: string,
+        environment?: string,
+        credentialType?: string,
+        clientId?: string,
+        realm?: string,
+        target?: string
+    ): Object {
```

Comment:
You can also do early return here as well. If any return false, you should return there instead of evaluating the rest. This is applicable in multiple where you are filtering.

Reply:
+1 for early return here, would make these operations less expensive in the long-run

Reply:
I will add early return for all the cases as needed.